src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -520,12 +519,22 @@ private BoundStatement RewriteStringForEachStatement(BoundForEachStatement node)
                 iterationVarDecl = new BoundExpressionStatement(assignment.Syntax, loweredAssignment);
                 RemovePlaceholderReplacement(deconstruction.TargetPlaceholder);
 
-                iterationVariables = assignment.LeftVariables.SelectAsArray(v => ((BoundLocal)v).LocalSymbol);
+                iterationVariables = GetLocalSymbols(assignment.LeftVariables);
             }
 
             return iterationVarDecl;
         }
 
+        private static ImmutableArray<LocalSymbol> GetLocalSymbols(ImmutableArray<BoundExpression> leftVariables)
+        {
+            var builder = ArrayBuilder<LocalSymbol>.GetInstance();
+            foreach (var variable in leftVariables.OfType<BoundLocal>())
```

Comment:
Change `OfType<BoundLocal>()` to a check for `Kind` inside the loop, to avoid allocations.